### PR TITLE
WIP: Dictionary Storage Backend

### PR DESCRIPTION
**Pull request by Jacob-Chmura**
Work in progress for _dictionary based_ `DGStorage` backend.

### CODE REVIEW

File: .pre-commit-config.yaml
Comment (Jacob-Chmura) [inline .pre-commit-config.yaml:1]: I will merge this upstream since it's annoying to deal with otherwise.

File: test/test_storage_impl.py
Comment (Jacob-Chmura) [inline test/test_storage_impl.py:1]: This is a very crude test suite design. I want to expand on this and properly test across a broader range of graphs to catch edge cases.

File: test/test_storage_impl.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+def test_update_multiple_events(DGStorageImpl):
+    events = []
+    storage = DGStorageImpl.from_events(events)
+    storage = storage.update([(1, 2, 3), (5, 10, 20)])
+    assert storage.to_events() == [(1, 2, 3), (5, 10, 20)]
+
+
+def test_get_dg_storage_backend():
+    backend = get_dg_storage_backend()
+    print(backend)
```
Comment (Jacob-Chmura) [inline test/test_storage_impl.py:111]:
``` suggestion
    assert get_dg_storage_backend() == DGStorageDictBackend
```

File: opendg/_storage/base.py [hunk 2]
```diff
@@ -30,14 +28,6 @@ def slice_nodes(self, nodes: List[int]) -> 'DGStorageBase':
     def get_nbrs(self, nodes: List[int]) -> Dict[int, List[Tuple[int, int]]]:
         r"""Return a list of neighbour, timestamp pairs for each node in the nodes list."""
 
-    @abstractmethod
```
Comment (Jacob-Chmura) [inline opendg/_storage/base.py:33]: It's not immediately clear that we will need these as part of the generic API, since _laziness_ is an implementation detail more than anything. For instance, the current naive dictionary implementation is automatically materialized. Hence, I removed these for the time being.

File: opendg/_storage/base.py [hunk 3]
```diff
@@ -48,14 +38,18 @@ def temporal_coarsening(
     ) -> 'DGStorageBase':
         r"""Re-index the temporal axis of the dynamic graph."""
 
+    def __len__(self) -> int:
```
Comment (Jacob-Chmura) [inline opendg/_storage/base.py:41]: Among other things, this seems relevant to have.

File: opendg/_storage/base.py [hunk 4]
```diff
@@ -48,14 +38,18 @@ def temporal_coarsening(
...
     @property
     @abstractmethod
-    def start_time(self) -> int:
+    def start_time(self) -> Optional[int]:
         r"""The start time of the temporal graph."""
 
     @property
     @abstractmethod
-    def end_time(self) -> int:
+    def end_time(self) -> Optional[int]:
```
Comment (Jacob-Chmura) [inline opendg/_storage/base.py:52]: We need to decide upon certain graph semantics. I'm assuming we allow empty graphs. In that case, start/end time are undefined (`None`).

File: opendg/_storage/base.py [hunk 5]
```diff
@@ -72,3 +66,9 @@ def num_edges(self) -> int:
     @abstractmethod
     def num_timestamps(self) -> int:
         r"""The total number of unique timestamps encountered over the temporal graph."""
+
+    def _check_slice_time_args(self, start_time: int, end_time: int) -> None:
```
Comment (Jacob-Chmura) [inline opendg/_storage/base.py:70]: Will continue to add in helper functions that will be common across backends into the base class (including device mixins, etc.)

File: opendg/_storage/backends/__init__.py [hunk 6]
```diff
@@ -0,0 +1,7 @@
+from opendg._storage.backends.dict_backend import DGStorageDictBackend
+
+DGStorageBackends = {
+    'DictionaryBackend': DGStorageDictBackend,
```
Comment (Jacob-Chmura) [inline opendg/_storage/backends/__init__.py:4]: Perhaps using the class name string itself is a bit better, I'm not entirely sure

File: opendg/_storage/backends/dict_backend.py
Comment (Jacob-Chmura) [inline opendg/_storage/backends/dict_backend.py:1]: This is a very naive and incomplete implementation. There are plenty of optimizations to be had, and some methods can be written a bit better. But this is the gist of it which should enable benchmarks / correctness testing.

File: opendg/_storage/backends/dict_backend.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        if not isinstance(events, list):
+            events = [events]
+        for t, u, v in events:
+            self._events_dict[t] = (u, v)
+        return self
+
+    def temporal_coarsening(
+        self, time_delta: int, agg_func: str = 'sum'
+    ) -> 'DGStorageBase':
+        raise NotImplementedError()
```
Comment (Jacob-Chmura) [inline opendg/_storage/backends/dict_backend.py:72]: TODO

File: opendg/_storage/backends/dict_backend.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        nbrs_dict = defaultdict(list)
+        for t, (u, v) in self._events_dict.items():
+            if u in nodes:
+                nbrs_dict[u].append((v, t))
+            if v in nodes:
+                nbrs_dict[v].append((u, t))
+
+        return dict(nbrs_dict)
+
+    def update(self, events: Union[Event, List[Event]]) -> 'DGStorageBase':
```
Comment (Jacob-Chmura) [inline opendg/_storage/backends/dict_backend.py:62]: Need to update cache

File: opendg/_storage/backends/dict_backend.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+from opendg.typing import Event
+
+from ..base import DGStorageBase
+
+
+class DGStorageDictBackend(DGStorageBase):
+    r"""Dictionary implementation of temporal graph storage engine."""
+
+    def __init__(self, events_dict: Dict[int, Tuple[int, int]]) -> None:
+        self._events_dict: Dict[int, Tuple[int, int]] = events_dict
```
Comment (Jacob-Chmura) [inline opendg/_storage/backends/dict_backend.py:13]: Values should be `List[Tuple[int, int]]`

File: opendg/_storage/backends/dict_backend.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+        events_dict = defaultdict(list)
+        for t, u, v in events:
+            events_dict[t].append((u, v))
+        return cls(events_dict)
+
+    def to_events(self) -> List[Event]:
+        events: List[Event] = []
+        for t, edges in self._events_dict.items():
+            for u, v in edges:
+                events.append((t, u, v))
```
Comment (shenyangHuang) [inline opendg/_storage/backends/dict_backend.py:33]: this appending will be extremely slow for large graphs, 
this is one of the reasons earlier TG implementations are very slow and can't work on large graphs as they append to large lists

File: opendg/_storage/backends/dict_backend.py [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+        self._end_time: Optional[int] = None
+        self._num_nodes: Optional[int] = None
+        self._num_edges: Optional[int] = None
+        self._num_timestamps: Optional[int] = None
+
+    @classmethod
+    def from_events(cls, events: List[Event]) -> 'DGStorageBase':
+        events_dict = defaultdict(list)
+        for t, u, v in events:
+            events_dict[t].append((u, v))
```
Comment (shenyangHuang) [inline opendg/_storage/backends/dict_backend.py:26]: this should be fine for the most part, maybe slow for coarse time granlarity as you are appending many edges to a single timestamp

File: opendg/_storage/backends/dict_backend.py [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+    ) -> 'DGStorageBase':
+        self._check_temporal_coarsening_args(time_delta, agg_func)
+
+        # This assert is here to make mypy happy
+        assert self.start_time is not None
+        assert self.end_time is not None
+        total_time = self.end_time - self.start_time
+
+        # TODO: Support other time_delta formats
+        interval_size = total_time // time_delta
```
Comment (shenyangHuang) [inline opendg/_storage/backends/dict_backend.py:84]: #15 we can implement some similar idea here

File: opendg/_storage/base.py (hunk 3 above)
Comment (shenyangHuang) [reply opendg/_storage/base.py:41]: Yes, this is the same as # of timestamps? or does this return some tensor sizes?

File: opendg/_storage/base.py [hunk 13]
```diff
@@ -48,27 +38,45 @@ def temporal_coarsening(
...
+
+    def __str__(self) -> str:
+        r"""Returns summary properties of the dynamic graph."""
+        return f'Dynamic Graph Storage Engine ({self.__class__.__name__}), Start Time: {self.start_time}, End Time: {self.end_time}, Nodes: {self.num_nodes}, Edges: {self.num_edges}, Timestamps: {self.num_timestamps}'
+
     @property
     @abstractmethod
-    def start_time(self) -> int:
-        r"""The start time of the temporal graph."""
+    def start_time(self) -> Optional[int]:
```
Comment (shenyangHuang) [inline opendg/_storage/base.py:51]: why is it optional?

File: test/test_storage_impl.py [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+    assert storage.end_time == 1
+    assert storage.num_nodes == 2
+    assert storage.num_edges == 1
+    assert storage.num_timestamps == 1
+
+
+def test_slice_time_bad_slice(DGStorageImpl):
+    events_dict = {1: (2, 3), 5: (10, 20)}
+    storage = DGStorageImpl(events_dict)
+    with pytest.raises(ValueError):
```
Comment (shenyangHuang) [inline test/test_storage_impl.py:53]: idea: maybe we should a custom class for errors? some errors would be invalid operation `InvalidOpsError` (like discretizing the graph into three snapshots when there are 2 edges only, need more thinking / testing). Others would be time related errors `ChronoError`, such as if you are slicing the timestamp with an interval either earlier than start time or later than end time?

File: test/test_storage_impl.py [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+    pass
+
+
+@pytest.mark.skip(reason='Not implemented')
+def test_temporal_coarsening_single_event_graph(DGStorageImpl):
+    pass
+
+
+@pytest.mark.skip(reason='Not implemented')
+def test_temporal_coarsening_causes_time_gap(DGStorageImpl):
```
Comment (shenyangHuang) [inline test/test_storage_impl.py:142]: these should have specific error message telling the user this is no valid because xxx

File: opendg/_storage/base.py (hunk 3 above)
Comment (Jacob-Chmura) [reply opendg/_storage/base.py:41]: Currently, I have implemented it as the # of timestamps, which I think is the most reasonable semantic for a dynamic graph _length_.

File: opendg/_storage/base.py (hunk 13 above)
Comment (Jacob-Chmura) [reply opendg/_storage/base.py:51]: Because the graph could be empty

File: test/test_storage_impl.py (hunk 14 above)
Comment (Jacob-Chmura) [reply test/test_storage_impl.py:53]: Yes, this is a good idea and we should do this whenever possible. For simplicity I've just raised ValueError's but having a custom error type for _time gaps_ makes sense. Generally though, I think using built in exception, where they make sense, is best (e.g. value error ["Raised when an operation or function receives an argument that has the right type but an inappropriate value"](https://docs.python.org/3/library/exceptions.html#ValueError))

File: opendg/_storage/backends/dict_backend.py (hunk 10 above)
Comment (Jacob-Chmura) [reply opendg/_storage/backends/dict_backend.py:33]: Can you elaborate? Perhaps I am missing something, but cpython lists are implemented as dynamic arrays. Therefore, appending to the end of a list has `O(1)` amortized time complexity (accounting for resizes), and has decent cache locality due to contiguous memory access (until the resize is triggered).

By your comment, do you simply mean that some appends will be much slower than others (when an array resize is triggered), or are you referring to the python loop which is, indeed very slow, but this does not really have much to do with the list appending?
